File: javascript/leetcode/spiral-matrix-iii.js
/**
 * @param {number} R
 * @param {number} C
 * @param {number} r0
 * @param {number} c0
 * @return {number[][]}
 */
var spiralMatrixIII = function(R, C, r0, c0) {
    var r = r0;
    var c = c0;
    var result = [[r, c]];
    var visits = 1;
    var steps = 1;
    while (visits < R * C) {
        // across to the right
        for (var i = 0; i < steps; i++) {
            c++;
            if (r < R && r >= 0 && c < C && c >= 0) {
                result.push([r, c]);
                visits++;
            }
        }
    
        // down
        for (var i = 0; i < steps; i++) {
            r++;
            if (r < R && r >= 0 && c < C && c >= 0) {
                result.push([r, c]);
                visits++;
            }
        }
        steps++;
        
        // across to the right
        for (var i = 0; i < steps; i++) {
            c--;
            if (r < R && r >= 0 && c < C && c >= 0) {
                result.push([r, c]);
                visits++;
            }
        }
        
        // up
        for (var i = 0; i < steps; i++) {
            r--;
            if (r < R && r >= 0 && c < C && c >= 0) {
                result.push([r, c]);
                visits++;
            }
        }
        steps++;
    }
    return result;
};

File: ruby/leetcode/spiral-matrix-iii.rb
# @param {Integer} r
# @param {Integer} c
# @param {Integer} r0
# @param {Integer} c0
# @return {Integer[][]}
def spiral_matrix_iii(r, c, r0, c0)
    rx = r0
    cx = c0
    result = [[rx, cx]]
    visits = 1
    steps = 1
    while visits < (r * c)
        # Go across to the right
        for i in 0...steps do
            cx += 1
            if rx < r && rx >= 0 && cx < c && cx >= 0
                result << [rx, cx]
                visits += 1
            end
        end

        # Go down
        for i in 0...steps do
            rx += 1
            if rx < r && rx >= 0 && cx < c && cx >= 0
                result << [rx, cx]
                visits += 1
            end
        end
        steps += 1

        # Go back to the left
        for i in 0...steps do
            cx -= 1
            if rx < r && rx >= 0 && cx < c && cx >= 0
                result << [rx, cx]
                visits += 1
            end
        end

        # Go up
        for i in 0...steps do
            rx -= 1
            if rx < r && rx >= 0 && cx < c && cx >= 0
                result << [rx, cx]
                visits += 1
            end
        end
        steps += 1
    end
    return result
end
